builtin/logical/rabbitmq/backend.go

```diff
@@ -0,0 +1,118 @@
+package rabbitmq
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"github.com/michaelklishin/rabbit-hole"
+)
+
+// Factory creates and configures Backends
+func Factory(conf *logical.BackendConfig) (logical.Backend, error) {
+	return Backend().Setup(conf)
+}
+
+// Backend creates a new Backend
+func Backend() *framework.Backend {
+	var b backend
+	b.Backend = &framework.Backend{
+		Help: strings.TrimSpace(backendHelp),
+
+		PathsSpecial: &logical.Paths{
+			Root: []string{
+				"config/*",
+			},
+		},
+
+		Paths: []*framework.Path{
+			pathConfigConnection(&b),
+			pathConfigLease(&b),
+			pathRoles(&b),
+			pathRoleCreate(&b),
+		},
+
+		Secrets: []*framework.Secret{
+			secretCreds(&b),
+		},
+
+		Clean: b.ResetClient,
+	}
+
+	return b.Backend
+}
+
+type backend struct {
+	*framework.Backend
+
+	client *rabbithole.Client
+	lock   sync.Mutex
+}
+
+// DB returns the database connection.
+func (b *backend) Client(s logical.Storage) (*rabbithole.Client, error) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// If we already have a client, we got it!
+	if b.client != nil {
+		return b.client, nil
+	}
+
+	// Otherwise, attempt to make connection
+	entry, err := s.Get("config/connection")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil,
+			fmt.Errorf("configure the client connection with config/connection first")
+	}
+
+	var connConfig connectionConfig
+	if err := entry.DecodeJSON(&connConfig); err != nil {
+		return nil, err
+	}
+
+	b.client, err = rabbithole.NewClient(connConfig.URI, connConfig.Username, connConfig.Password)
+	if err != nil {
+		return nil, err
+	}
+
+	return b.client, nil
+}
+
+// ResetClient forces a connection next time Client() is called.
+func (b *backend) ResetClient() {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
```

Comment:
@KevinPike Regarding `rabbithole.Client`'s `Close` method, I saw this: https://godoc.org/github.com/michaelklishin/rabbit-hole#Client.CloseConnection. Maybe it is worth checking out!

Regarding deleting users/policies/vhosts, I only wanted your second look at it, just in case it was missed out. Usually, it should be taken care of in the secrets revocation scenario, rather than here, which I see that I have commented on already anyways. So, I guess this is good enough.

Just check the `Close` on client.